### PR TITLE
[ENH] Implementation of BIDS utilities using pybids

### PR DESCRIPTION
**Pull request by oesteban**
- [x] Update to pybids>=0.7
- [x] Use pybids in `niworkflows.utils.bids`
- [x] Use pybids in `niworkflows.interfaces.bids`
- [x] Add doctests to `ReadSidecarJSON`
- [x] Initialize BIDSLayout within interfaces/functions if none is previously given but a `bids_dir` is available.

### CODE REVIEW

**Comment by effigies** [comment]
Before wandering too far down this path, I would go ahead and bump up to `pybids>=0.7`.

**Comment by effigies** [comment]
Won't be able to review until later tonight at the earliest.

**Comment by oesteban** [comment]
Of course, no rush. It can be done on Monday, no need to work over the weekend.

File: niworkflows/utils/bids.py [hunk 1]
```diff
@@ -163,8 +168,10 @@ def collect_data(dataset, participant_label, task=None, echo=None):
     if echo:
         queries['bold']['echo'] = echo
 
-    subj_data = {modality: [x.filename for x in layout.get(**query)]
-                 for modality, query in queries.items()}
+    subj_data = {
+        dtype: sorted([x.path for x in layout.get(subject=participant_label,
```
Comment (effigies) [inline niworkflows/utils/bids.py]:
```suggestion
        dtype: sorted(layout.get(return_type='file', subject=participant_label,
                                 extensions=['nii', 'nii.gz'], **query))
```

File: niworkflows/utils/bids.py [hunk 2]
```diff
@@ -163,8 +168,10 @@ def collect_data(dataset, participant_label, task=None, echo=None):
...
+                                                  extensions=['nii', 'nii.gz'], **query)])
```
Comment (effigies) [inline niworkflows/utils/bids.py]:
```suggestion
```

File: niworkflows/interfaces/bids.py [hunk 3]
```diff
@@ -293,46 +337,86 @@ def _run_interface(self, runtime):
...
-            if id_value is not None:
-                self._results[key] = outputs.get(key)
-
+        self.layout = self.inputs.bids_dir or self.layout
+        self.layout = _init_layout(self.inputs.in_file,
+                                   self.layout,
+                                   self.inputs.bids_validate)
+
+        # Fill in BIDS entities of the output ("*_id")
+        output_keys = [key for key in list(BIDSInfoOutputSpec().get().keys())]
```
Comment (effigies) [inline niworkflows/interfaces/bids.py]:
```suggestion
        output_keys = list(BIDSInfoOutputSpec().get().keys())
```

File: niworkflows/__about__.py [hunk 4]
```diff
@@ -52,14 +52,14 @@
 ]
 
 REQUIRES = [
-    'grabbit==0.2.3',
+    'grabbit==0.2.6',
```
[the remark this answers is not in the document]
Comment (oesteban) [reply niworkflows/__about__.py]: :+1:

File: niworkflows/interfaces/bids.py [hunk 5]
```diff
@@ -8,71 +8,115 @@
...
 
 
-class BIDSInfoOutputSpec(TraitedSpec):
-    subject_id = traits.Str()
-    session_id = traits.Str()
-    task_id = traits.Str()
-    acq_id = traits.Str()
-    rec_id = traits.Str()
-    run_id = traits.Str()
+class BIDSInfoOutputSpec(DynamicTraitedSpec):
```
[the remark this answers is not in the document]
Comment (oesteban) [reply niworkflows/interfaces/bids.py]: We need this to be dynamic for the ReadSidecarJSON interface (I could make it a mixin, yes, but thought this is more readable). Additionally, in my mind would prefer that all outputs except for the subject id and suffix were dynamic, and define them only when they are set.

File: niworkflows/__about__.py [hunk 6]
```diff
@@ -52,14 +52,14 @@
 ]
 
 REQUIRES = [
-    'grabbit==0.2.3',
+    'grabbit',
```
Comment (effigies) [inline niworkflows/__about__.py]:
```suggestion
```

**Comment by oesteban** [comment]
Thumbs up?

**Comment by effigies** [comment]
What's up with https://circleci.com/gh/oesteban/workflows/fmriprep/tree/rf%2Fniworkflows-pr-299?

**Comment by effigies** [comment]
`traits.trait_errors.TraitError: The trait 'sbref_file' of an EstimateReferenceImageInputSpec instance is an existing file name, but the path  'sub-100185_task-machinegame_run-01_sbref.nii.gz' does not exist.

**Comment by oesteban** [comment]
I'll fix that up before merging :+1: . After that, you'd be okay?

**Comment by effigies** [comment]
Yeah, this looks fine.

**Comment by oesteban** [comment]
Okay, I'm pretty positive this will do: https://github.com/oesteban/fmriprep/commit/8e1879b549e95b01a0eac74658a64a0463b34e8f

Let's wait and see if the assumption is correct (https://circleci.com/workflow-run/5ecd01fd-11a1-4328-9860-641c4b8fc318)

**Comment by effigies** [comment]
No, it's actually that it should be

```Diff
- sbref_file = files[0].filename
+ sbref_file = files[0].path
```

on L276.

**Comment by effigies** [comment]
Or you could keep the `return_type='file'` and just use `sbref_file = files[0]`.

**Comment by oesteban** [comment]
Right :+1:

**Pull request merged**